### PR TITLE
DOC-3152: add a dialect reference page to redis.io

### PR DESCRIPTION
**Pull request by dwdougherty**
Documentation change only: adding a new dialect reference page for redis.io.

@adrianoamaral Please review, approve, and merge.

### CODE REVIEW

File: docs/docs/advanced-concepts/dialects.md [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+    `@name:James Brown`
+
+    Here, the field modifier `@name` is followed by two words, `James` and `Brown`.
+
+    In `DIALECT 1`, this query would be interpreted as find `James Brown` in the `@name` field.
+    In `DIALECT 2`, this query would be interpreted as find `James` in the `@name` field, and `Brown` in any text field. In other words, it would be interpreted as `(@name:James) Brown`. 
+    In `DIALECT 2`, you could achieve the dialect 1 behavior by updating your query to `@name:(James Brown)`.
+
+1. Your query uses `"..."`, `~`, `-`, and/or `%`. Consider a simple query with negation:
```
Comment (adrianoamaral) [inline docs/docs/advanced-concepts/dialects.md:42]:
```suggestion
2. Your query uses `"..."`, `~`, `-`, and/or `%`. Consider a simple query with negation:
```

File: docs/docs/advanced-concepts/dialects.md [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+    `-hello world`
+
+    In `DIALECT 1`, this query is interpreted as find values in any field that do not contain `hello` and do not contain `world`; the equivalent of `-(hello world)` or `-hello -world`.
+    In `DIALECT 2`, this query is interpreted as `-hello` and `world` (only `hello` is negated).
+    In `DIALECT 2`, you could achieve the dialect 1 behavior by updating your query to `-(hello world)`.
+
+1. Your query used `|`. Consider the simple query:
```
Comment (adrianoamaral) [inline docs/docs/advanced-concepts/dialects.md:50]:
```suggestion
3. Your query used `|`. Consider the simple query:
```

File: docs/docs/advanced-concepts/dialects.md [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+    `hello world | "goodbye" moon`
+
+    In `DIALECT 1`, this query is interpreted as searching for `(hello world | "goodbye") moon`.
+    In `DIALECT 2`, this query is interpreted as searching for either `hello world` `"goodbye" moon`.
+
+1. Your query uses a wildcard pattern. Consider the simple query:
```
Comment (adrianoamaral) [inline docs/docs/advanced-concepts/dialects.md:57]:
```suggestion
4. Your query uses a wildcard pattern. Consider the simple query:
```

**Review by adrianoamaral — COMMENTED** [review]
Few comments and suggestions

File: docs/docs/advanced-concepts/dialects.md [hunk 4]
```diff
@@ -0,0 +1,158 @@
+---
+title: "Query dialects"
+linkTitle: "Query dialects"
+weight: 5
+description: >
+    Learn how to use query dialects
+---
+
+Redis Stack currently supports four query dialects for use with the `FT.SEARCH`, `FT.AGGREGATE`, and other search and query commands.
+
```
[the remark this answers is not in the document]
Comment (K-Jo) [reply docs/docs/advanced-concepts/dialects.md:10]: search without capital S

[the remark this answers is not in the document]
Comment (K-Jo) [reply docs/docs/advanced-concepts/dialects.md:10]: I think this is even a bit hiding it no @adrianoamaral?  Why not just say that we didn't want to break the API for existing users while enhancing it for enabling new use cases

>  do not need to undergo extensive modifications.

[the remark this answers is not in the document]
Comment (adrianoamaral) [reply docs/docs/advanced-concepts/dialects.md:10]: Solved! Tks

File: docs/docs/advanced-concepts/dialects.md [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+`FT.SEARCH ... DIALECT 4`
+
+Dialect version 4 will improve performance in four different scenarios:
+
+1. Skip sorter - applied when there is no sorting to be done. The query can return once it reaches the `LIMIT` of requested results.
+1. Partial range - applied when there is a `SORTBY` on a numeric field, either with no filter or with a filter by the same numeric field. Such queries will iterate on a range large enough to satisfy the `LIMIT` of requested results.
+1. Hybrid - applied when there is a `SORTBY` on a numeric field in addition to another non-numeric filter. It could be the case that some results will get filtered, leaving too small a range to satisfy any specified `LIMIT`. In such cases, the iterator then is re-wound and additional iterations occur to collect result up to the requested `LIMIT`.
+1. No optimization - If there is a sort by score or by a non-numeric field, there is no other option but to retrieve all results and compare their values to the search parameters.
+
```
Comment (dwdougherty) [inline docs/docs/advanced-concepts/dialects.md:145]: ## Use `FT.EXPLAINCLI` to compare dialects
	
The [`FT.EXPLAINCLI`](https://redis.io/commands/ft.explaincli/) is a powerful tool that provides a window into the inner workings of your queries. It's like a roadmap that details your query's journey from start to finish.

When you run `FT.EXPLAINCLI`, it returns an array representing the execution plan of a complex query. This plan is a step-by-step guide of how Redis interprets your query and how it plans to fetch results. It's a behind-the-scenes look at the process, giving you insights into how the search engine works.

The `FT.EXPLAINCLI` accepts a `DIALECT` argument, allowing you to execute the query using different dialect versions, allowing you to compare the resulting query plans.

To use `FT.EXPLAINCLI`, you need to provide an index and a query predicate. The index is the name of the index you created using `FT.CREATE`, and the query predicate is the same as if you were sending it to `FT.SEARCH` or `FT.AGGREGATE`.

Here's an example of how to use `FT.EXPLAINCLI` to understand differences in dialect versions 1 and 2.

Negation of the intersection between tokens `hello` and `world`:

```FT.EXPLAINCLI idx:dialects "-hello world" DIALECT 1
1) NOT {
2)   INTERSECT {
3)     hello
4)     world
5)   }
6) }
7)
```

Intersection of the negation of the token `hello` together with token `world`:

```
FT.EXPLAINCLI idx:dialects "-hello world" DIALECT 2
 1) INTERSECT {
 2)   NOT {
 3)     hello
 4)   }
 5)   UNION {
 6)     world
 7)     +world(expanded)
 8)   }
 9) }
10) 
```

Same result as `DIALECT 1`:

```
FT.EXPLAINCLI idx:dialects "-(hello world)" DIALECT 2
1) NOT {
2)   INTERSECT {
3)     hello
4)     world
5)   }
6) }
7) 
```

{{% alert title=Note %}}
`FT.EXPLAIN` doesn't execute the query. It only explains the plan. It's a way to understand how your query is interpreted by the query engine, which can be invaluable when you're trying to optimize your searches.
{{% /alert %}}

File: docs/docs/advanced-concepts/dialects.md [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+`FT.SEARCH ... DIALECT 4`
+
+Dialect version 4 will improve performance in four different scenarios:
+
+1. **Skip sorter** - applied when there is no sorting to be done. The query can return once it reaches the `LIMIT` of requested results.
+1. **Partial range** - applied when there is a `SORTBY` on a numeric field, either with no filter or with a filter by the same numeric field. Such queries will iterate on a range large enough to satisfy the `LIMIT` of requested results.
+1. **Hybrid** - applied when there is a `SORTBY` on a numeric field in addition to another non-numeric filter. It could be the case that some results will get filtered, leaving too small a range to satisfy any specified `LIMIT`. In such cases, the iterator then is re-wound and additional iterations occur to collect result up to the requested `LIMIT`.
+1. **No optimization** - If there is a sort by score or by a non-numeric field, there is no other option but to retrieve all results and compare their values to the search parameters.
+
```
Comment (dwdougherty) [inline docs/docs/advanced-concepts/dialects.md:148]:
```suggestion
## Use `FT.EXPLAINCLI` to compare dialects
	
The [`FT.EXPLAINCLI`](https://redis.io/commands/ft.explaincli/) is a powerful tool that provides a window into the inner workings of your queries. It's like a roadmap that details your query's journey from start to finish.

When you run `FT.EXPLAINCLI`, it returns an array representing the execution plan of a complex query. This plan is a step-by-step guide of how Redis interprets your query and how it plans to fetch results. It's a behind-the-scenes look at the process, giving you insights into how the search engine works.

The `FT.EXPLAINCLI` accepts a `DIALECT` argument, allowing you to execute the query using different dialect versions, allowing you to compare the resulting query plans.

To use `FT.EXPLAINCLI`, you need to provide an index and a query predicate. The index is the name of the index you created using `FT.CREATE`, and the query predicate is the same as if you were sending it to `FT.SEARCH` or `FT.AGGREGATE`.

Here's an example of how to use `FT.EXPLAINCLI` to understand differences in dialect versions 1 and 2.

Negation of the intersection between tokens `hello` and `world`:

```FT.EXPLAINCLI idx:dialects "-hello world" DIALECT 1
1) NOT {
2)   INTERSECT {
3)     hello
4)     world
5)   }
6) }
7)
```

Intersection of the negation of the token `hello` together with token `world`:

```
FT.EXPLAINCLI idx:dialects "-hello world" DIALECT 2
 1) INTERSECT {
 2)   NOT {
 3)     hello
 4)   }
 5)   UNION {
 6)     world
 7)     +world(expanded)
 8)   }
 9) }
10) 
```

Same result as `DIALECT 1`:

```
FT.EXPLAINCLI idx:dialects "-(hello world)" DIALECT 2
1) NOT {
2)   INTERSECT {
3)     hello
4)     world
5)   }
6) }
7) 
```

{{% alert title=Note %}}
`FT.EXPLAIN` doesn't execute the query. It only explains the plan. It's a way to understand how your query is interpreted by the query engine, which can be invaluable when you're trying to optimize your searches.
{{% /alert %}}

```

Comment (dwdougherty) [inline docs/docs/advanced-concepts/dialects.md:148]:
````suggestion
## Use `FT.EXPLAINCLI` to compare dialects
	
The [`FT.EXPLAINCLI`](https://redis.io/commands/ft.explaincli/) is a powerful tool that provides a window into the inner workings of your queries. It's like a roadmap that details your query's journey from start to finish.

When you run `FT.EXPLAINCLI`, it returns an array representing the execution plan of a complex query. This plan is a step-by-step guide of how Redis interprets your query and how it plans to fetch results. It's a behind-the-scenes look at the process, giving you insights into how the search engine works.

The `FT.EXPLAINCLI` accepts a `DIALECT` argument, allowing you to execute the query using different dialect versions, allowing you to compare the resulting query plans.

To use `FT.EXPLAINCLI`, you need to provide an index and a query predicate. The index is the name of the index you created using `FT.CREATE`, and the query predicate is the same as if you were sending it to `FT.SEARCH` or `FT.AGGREGATE`.

Here's an example of how to use `FT.EXPLAINCLI` to understand differences in dialect versions 1 and 2.

Negation of the intersection between tokens `hello` and `world`:

```FT.EXPLAINCLI idx:dialects "-hello world" DIALECT 1
1) NOT {
2)   INTERSECT {
3)     hello
4)     world
5)   }
6) }
7)
```

Intersection of the negation of the token `hello` together with token `world`:

```
FT.EXPLAINCLI idx:dialects "-hello world" DIALECT 2
 1) INTERSECT {
 2)   NOT {
 3)     hello
 4)   }
 5)   UNION {
 6)     world
 7)     +world(expanded)
 8)   }
 9) }
10) 
```

Same result as `DIALECT 1`:

```
FT.EXPLAINCLI idx:dialects "-(hello world)" DIALECT 2
1) NOT {
2)   INTERSECT {
3)     hello
4)     world
5)   }
6) }
7) 
```

{{% alert title=Note %}}
`FT.EXPLAIN` doesn't execute the query. It only explains the plan. It's a way to understand how your query is interpreted by the query engine, which can be invaluable when you're trying to optimize your searches.
{{% /alert %}}

````

**Comment by dwdougherty** [comment]
@adriano Please re-review.

**Review by adrianoamaral — APPROVED** [review]

**Pull request merged**